### PR TITLE
Add 4.0 sets

### PR DESCRIPTION
**Pull request by imring**
Add `Marechaussee Hunter` and `Golden Troupe`.

### CODE REVIEW

File: pkg/shortcut/artifacts.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ var SetNameToKey = map[string]keys.Set{
 	"gildeddreams":            keys.GildedDreams,
 	"gilded":                  keys.GildedDreams,
 	"gd":                      keys.GildedDreams,
+	"goldentroupe":            keys.GoldenTroupe,
```
Comment (k0l11) [inline pkg/shortcut/artifacts.go:48]: Please add "gt" as an alias.

File: internal/artifacts/goldentroupe/goldentroupe.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	s.core.Log.NewEvent("golden troupe 4pc proc'd", glog.LogArtifactEvent, s.char.Index)
+}
+
+func (s *Set) clearBuff(src int) func() {
+	return func() {
+		if s.lastSwap != src {
+			return
+		}
+
+		if s.core.Player.Active() == s.char.Index {
```
Comment (k0l11) [inline internal/artifacts/goldentroupe/goldentroupe.go:104]: Please invert the condition to perform an early return to reduce nesting.

File: internal/artifacts/goldentroupe/goldentroupe.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	// This effect will be cleared 2s after taking the field.
+	if count >= 4 {
+		s.buff = make([]float64, attributes.EndStatType)
+		s.buff[attributes.DmgP] = 0.25
+
+		c.Events.Subscribe(event.OnCharacterSwap, func(args ...interface{}) bool {
+			prev := args[0].(int)
+			next := args[1].(int)
+			if prev == char.Index {
+				s.gainBuff()
```
Comment (k0l11) [inline internal/artifacts/goldentroupe/goldentroupe.go:71]: I think this part is bugged, because you never invalidate the s.clearBuff buff task when gaining the buff. Imagine the char starts off-field and you swap to them. Then s.lastSwap is set and s.clearBuff is added as a task. When swapping off again, s.lastSwap does not change so the s.clearBuff will still execute even though the char is off-field again. Please check if this is the case and fix it.

TLDR:
```
s.gainBuff()
s.lastSwap = -1
```

File: internal/artifacts/marechausseehunter/marechausseehunter.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+	// When current HP increases or decreases, CRIT Rate will be increased by 12% for 5s. Max 3 stacks.
+	if count >= 4 {
+		s.buff = make([]float64, attributes.EndStatType)
+
+		c.Events.Subscribe(event.OnPlayerHPDrain, func(args ...interface{}) bool {
+			di := args[0].(player.DrainInfo)
+			if c.Player.Active() != char.Index {
+				return false
+			}
```
[the remark this answers is not in the document]
Comment (imring) [reply internal/artifacts/marechausseehunter/marechausseehunter.go:82]: but it's not in the description

File: internal/artifacts/marechausseehunter/marechausseehunter.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+			s.onChangeHP()
+			return false
+		}, fmt.Sprintf("mh-4pc-%v", char.Base.Key.String()))
+
+		c.Events.Subscribe(event.OnHeal, func(args ...interface{}) bool {
+			index := args[1].(int)
+			amount := args[2].(int)
+			if c.Player.Active() != char.Index {
+				return false
+			}
```
[the remark this answers is not in the document]
Comment (k0l11) [reply internal/artifacts/marechausseehunter/marechausseehunter.go:99]: Nevermind.

File: internal/artifacts/marechausseehunter/marechausseehunter.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (k0l11) [reply internal/artifacts/marechausseehunter/marechausseehunter.go:82]: Nevermind.

**Review by k0l11 — APPROVED** [review]
